### DISCUSSION TITLE
Presentation and demo of Joyride, an extension for hacking VS Code in user space (like Emacs)

### DISCUSSION THREAD

**Discussion by PEZ** [microsoft/vscode-discussions#281] [Extension Show and Tell]
[Joyride](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.joyride) is an extension for Visual Studio Code that makes the editor scriptable in user space. Joyride also gives the user a REPL where the extensible parts of VS Code, and the user's Joyride scripts, can be modified while they are running, without any need for reloading the extension host. Both of these aspects are quite similar to how things work for Emacs users.

Here's a recording of a presentation and demo of Joyride that I did yesterday together with the co-creator of the extension:
* https://www.youtube.com/watch?v=ObjIR08t3lg

The presentation presents the Why we created Joyride, what it is, and what users can do with it. It is also a demo of some of the things that it can be used for. At a few places it is also a demo of what Interactive Programming (aka REPL driven development) brings to the table. (The part with modifying scripts and functions while they are running, without reloading the window.)

The audience was mostly consisting users of editors, and not extension developers. With the aim of inspiring them to start extending VS Code for themselves.

I've created a repository with the slides and the examples. Including instructions for some of the examples.

* https://github.com/PEZ/london-clojurians-joyride

It would be totally awesome with some feedback from other extension authors, and people interested in VS Code extensibility, around what is demonstrated in the recording! ❤️ 🙏

**Comment by sinewalker** [comment]
This will lower my barrier to hacking on Code, on two fronts:

- It exposes all the VSCode APIs in a REPL: I don't have to learn how to make an extension first, before hacking on Code
- It's using Clojure(script?) rather than Typescript, so it feels a lot like elisp

Thank you for making this! I've been hovering on the edge of playing with Clojure for a long while, and thinking "if only there were a way to hack Code like you can Emacs, and it used Clojure for the language... that'd be so nice, but also so niche it'll never happen..."  😺 

I had to drop this note in here before I finished watching your presentation. I'll go watch the rest now. Also I'm going to go read [Clojure for the Brave and True](https://www.braveclojure.com/#quests) with an _actual use in mind_.

**Reply by PEZ** [reply]
So great to hear! And very cool that you pick the role of the REPL here. It is a key ingredient and what makes it worthy to compare to Emacs scripting. 

Here's a template for using Joyride for Advent of code, which might be a fun way to crack open the Clojure box:

* https://github.com/PEZ/joyride-aoc-2022

Also, Calva has a command for introducing how it leverages the REPL and also for introducing Clojure: **Calva: Fire up The Getting Started REPL**.
